### PR TITLE
Fix a bug with foreach intents when `--baseline` is used

### PR DESCRIPTION
**Pull request by e-kayrakli**
Follow on to https://github.com/chapel-lang/chapel/pull/24451

We've encountered this before with a test that used `--baseline` and assumed it was specific to the test, but it looks like the issue was widespread. The reproducer is (based on @stonea's in https://github.com/Cray/chapel-private/issues/5816#issuecomment-1944843837):

```chpl
record Inner {
  var innerField = 42;
}

record Outer {
  var outerField : Inner;

  proc init() {
    outerField = new Inner();
  }

  iter const myIter() {
    foreach idx in 0..10 /* with (ref this)*/ do  // Uncomment with clause and it'll work
      if outerField.innerField == 42 then yield outerField.innerField;
  }
}

var x = new Outer();

for i in x.myIter() do writeln(i);
```

Fortunately, it is pretty esoteric to `--baseline`, I think. The more concerning issue is about the compiler implementation of intents in my view, where generating AST for implicit and explicit intents go through very different code paths. But that's a big overhaul.

The problem with the bug has to do with how shadow variables and the jumps interact within the `advance` function. The following is a sketch of the generated C code.

```c
void advance(myType) {
  int* fieldRef = NULL;

  if iteratorHasTrippedOnceAlready {
    goto X;
  }

  fieldRef = &myType.field

  for (....) {
    int cur = *fieldRef;
X:
    // something something
  }
}
```

In an uninlined iterator, `zip1` calls `advance`, which will _not_ skip over `fieldRef`'s def. However, when `zip3` calls advance (which happens in the second trip of the loop), we'll hit the goto that'll jump over the def. That will result in a null dereference when the for loop in `advance` begins its first full iteration after the goto jumps right in the middle of the loop body.

The solution here does what would happen with an explicit `ref` intent in the iterator in question. That is `fieldRef`'s def is moved inside the loop body. I think a better solution would be to put shadow variable initialization before the gotos, but I don't know if that would impact something else.

Another general worry is that we have been probably working with the assumption that iterators can't have loops with intents as you can't use a `forall` inside an iterator (?). However, with foreach intents that's no longer true. I think it'd be nice to revert that assumption to avoid ad-hoc workarounds.

Test:
- [x] linux64
- [x] linux64 `--baseline`

### CODE REVIEW

File: compiler/resolution/implementForallIntents.cpp [hunk 1]
```diff
@@ -1008,6 +1019,12 @@ static ShadowVarSymbol* createSVforFieldAccess(LoopWithShadowVarsInterface* fs,
   svar->type = svarType;
   fs->shadowVariables().insertAtTail(new DefExpr(svar));
   handleOneShadowVar(fs, svar);
+  //if (field->getModule()->modTag == MOD_USER) {
```
Comment (stonea) [inline compiler/resolution/implementForallIntents.cpp:1022]: Cruft?

File: compiler/resolution/implementForallIntents.cpp [hunk 2]
```diff
@@ -994,7 +995,17 @@ static ShadowVarSymbol* createSVforFieldAccess(LoopWithShadowVarsInterface* fs,
                                                Symbol* field)
 {
   bool isConst = ovar->isConstant() || field->isConstant();
-  VarSymbol* fieldRef = createFieldRef(fs->asExpr(), ovar, field, isConst);
+
+  Expr* anchor = nullptr;
```
Comment (stonea) [inline compiler/resolution/implementForallIntents.cpp:999]: Could you add a comment explaining what's happening here. (or at least indicate that it's a patch to workaround an issue that comes up when using `--baseline`.

File: compiler/resolution/implementForallIntents.cpp [hunk 3]
```diff
@@ -25,6 +25,7 @@
 #include "resolveIntents.h"
 #include "stringutil.h"
 #include "stmt.h"
+#include "view.h"
```
Comment (stonea) [inline compiler/resolution/implementForallIntents.cpp:28]: Cruft?

File: modules/standard/Set.chpl [hunk 4]
```diff
@@ -516,7 +516,8 @@ module Set {
       :yields: A constant reference to an element in this set.
     */
     iter const these() const ref {
-      foreach idx in 0..#_htb.tableSize with (const ref this) {
+      /*foreach idx in 0..#_htb.tableSize with (const ref this) {*/
```
Comment (stonea) [inline modules/standard/Set.chpl:519]: Remove commented version?

**Review by stonea — APPROVED** [review]

File: compiler/resolution/implementForallIntents.cpp (hunk 2 above)
Comment (e-kayrakli) [reply compiler/resolution/implementForallIntents.cpp:1002]: https://github.com/chapel-lang/chapel/pull/24556/commits/ae5576b4afda65dd0954e46786d5d044e109c44b

**Pull request merged**